### ISSUE TITLE
Constrained Optimization (especially lower+upper bounds)

### ISSUE THREAD

**Issue by JulianCologne** [gonum/gonum#1725]
Hello,

Coming from Python (scipy.optimize.minimize) and MATLAB (fmincon), I can't find a way to specify bounds (and other constraints), which to me are very essential parts of real world application optimization problem.

Python / Scipy:
```python
res = minimize(opti_fun, x0, bounds=mybounds, constraints=myconstraints)
```

Matlab:
```matlab
res = fmincon(opti_fun, x0, A, b, Aeq, beq, lowerbound, upperbound)
```

Looking through the documentation, I could not find that this is implemented.
Is this something that is being worked on? Or is it planned? Or is it too complicated / time consuming at the moment? Or did I just miss something here?

I would love to see this in go!

**Comment by vladimir-ch** [comment]
> I would love to see this in go!

Yes, we'd love that too. I'm aware that most practical optimization problems are constrained. Unfortunately, implementing a constrained solver is a big task that requires concentrated effort and our manpower and free time are both severely limited. I don't see it being added soon but we will surely be happy to welcome any help ;-)

Maybe L-BFGS-B would be a good starting point, it's an extension of a solver that we already have, so the task is potentially not so daunting but already a big step forward.

**Comment by JulianCologne** [comment]
Thanks for the response.
You are absolutely right, this is a big job. Maybe I find some time in the future and can contribute to it. Good tip with L-BFGS-B, I'll take a look at that. Thanks!

**Comment by sbinet** [comment]
as I noted in https://github.com/gonum/gonum/issues/354#issuecomment-395685405, a stop-gap solution (that has been used for decades in high energy physics) is to apply a parameter change that take into account the bounds:

```
P_internal = arcsin (2 * (P_external - a) / (b - a) - 1)

P_external = a + (b-a)/2 (sin P_internal + 1)
```

- https://cds.cern.ch/record/2296388/files/minuit.pdf (page 2)

**Comment by JulianCologne** [comment]
@sbinet Thanks for the suggestion.
Looks very interesting, I will try that! ;)

**Comment by crhntr** [comment]
I read the [Nelder Mead implementation in nlopt and noticed how it handles bounds](https://github.com/stevengj/nlopt/blob/d676b27aecb4f43ea4766af3c0c810dd56622953/src/algs/neldermead/nldrmd.c#L52-L77). Here is a branch with what I think it might look like added to the gonum implementation.
https://github.com/crhntr/gonum/tree/add-bounds-to-neldermead
